flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java

```diff
@@ -0,0 +1,263 @@
+        this.keyedStateBackend = keyedStateBackend;
+        this.stateChangelogWriter = stateChangelogWriter;
+        this.asyncExceptionHandler = asyncExceptionHandler;
+        this.periodicExecutor =
+                Executors.newSingleThreadScheduledExecutor(
+                        new ExecutorThreadFactory("periodic-materialization"));
+        this.streamFactory = shared -> checkpointStorageWorkerView.createTaskOwnedStateStream();
+        this.allowedNumberOfFailures = allowedNumberOfFailures;
+        this.materializationOnGoing = new AtomicBoolean(false);
+        this.retries = new AtomicInteger(allowedNumberOfFailures);
+
+        this.materializedId = 0;
+        this.materializedState = materializedState;
+
+        if (materializationEnabled) {
+            this.periodicExecutor.scheduleAtFixedRate(
+                    this::triggerMaterialization,
+                    periodicMaterializeInitDelay,
+                    periodicMaterializeInterval,
+                    TimeUnit.MILLISECONDS);
+        }
+    }
+
+    @VisibleForTesting
+    public void triggerMaterialization() {
+        mailboxExecutor.execute(
+                () -> {
+                    // Only one materialization ongoing at a time
+                    if (!materializationOnGoing.compareAndSet(false, true)) {
+                        return;
+                    }
+
+                    SequenceNumber upTo = stateChangelogWriter.lastAppendedSequenceNumber().next();
+
+                    if (upTo.equals(materializedState.lastMaterializedTo())) {
+                        return;
+                    }
+
+                    RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture =
+                            keyedStateBackend.snapshot(
+                                    materializedId++, // TO DISCUSS: seems this does not matter,
+                                    // based on
+                                    System.currentTimeMillis(),
+                                    streamFactory,
+                                    CHECKPOINT_OPTIONS);
+
+                    // TODO: add metadata to log FLINK-23170
+                    asyncOperationsThreadPool.execute(
+                            () -> asyncMaterialization(materializedRunnableFuture, upTo));
+                },
+                "materialization");
+    }
+
+    private void asyncMaterialization(
+            RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture,
+            SequenceNumber upTo) {
+
+        FileSystemSafetyNet.initializeSafetyNetForThread();
+        try {
+            FutureUtils.runIfNotDoneAndGet(materializedRunnableFuture);
+            mailboxExecutor.execute(
+                    () -> {
+                        materializedState =
+                                new MaterializedState(
+                                        getMaterializedResult(materializedRunnableFuture.get()),
```

Comment:
I think it's safer to call `future.get` from the async thread (even if it's a no-op currently).